### ISSUE TITLE
Can you add back the ability to train by specifying the max number of steps

### ISSUE THREAD

**Issue by ekeric13** [d8ahazard/sd_dreambooth_extension#703]
It seems that there was a major UI overhaul and some features got cut that I found pretty useful.

1. Ability to train by number of steps (currently just supports the number of steps per image)
2. Ability to save previews and checkpoints every X steps (currently just supports saving of per epoch)

This seems to be the current UI:
<img width="420" alt="Screenshot 2023-01-05 at 3 39 47 PM" src="https://user-images.githubusercontent.com/6489651/210900628-d6652240-b29a-4ef0-99f8-ba245d2c7311.png">

While this was the previous one where you had the option to go by total steps or by epochs:
<img width="416" alt="Screenshot 2023-01-05 at 3 40 02 PM" src="https://user-images.githubusercontent.com/6489651/210900670-f155a58f-5fb0-49be-adca-e65239915bbf.png">

Additionally I am confused by what "save model frequency by epoch" computes to.
If the value is 1 and the steps per epoch is 100 and you have 10 images, I would think that creates a model every 100 steps.
If the value is 2 I would think that creates a model every 200 steps.
etc.

Is that correct?

Anyway it would be great if you could add those options back in.

**Comment by ekeric13** [comment]
Another regression is no longer showing you how many images it has generated during regularization

<img width="468" alt="Screenshot 2023-01-05 at 4 12 00 PM" src="https://user-images.githubusercontent.com/6489651/210903836-1309a038-ae65-4697-995f-dec0a4bc1724.png">

**Comment by d8ahazard** [comment]
No, I have no plans on re-implementing anything based on the number of steps, and here's why:

For starters, and foremost...what does "number of steps" actually represent? In the original dreambooth script, it was a count of the times optimizer.step() was called. This doesn't account for batch size, which means a "step" could actually indicate anywhere from 1-10 iterations over images. Additionally, there's gradient accumulation steps, which, if accounted for, act as another multiplier on the "total number of steps" that one could consider the model having learned.

Regardless, this random ephemeral number has somehow made it's way into the standard lexicon when talking about training as some kind of tangible unit of measuring training depth. It's odd that it ever happened in the first place, but just because it became a standard way of talking about things doesn't mean it should continue. 


Additionally, it makes *everything* so much more complicated to talk in sense of steps, when what we're doing to get these steps is multiplying our number of training images by some number "steps per image", and then using that to accomplish the same thing.

But, if we just consider what we're actually trying to do, I'd posit that using epochs to describe and quantify training is so much better. "I'm having trouble training this subject, I already did 30k steps and it's still not great." means little without also knowing how many instance images were used. "I'm having trouble...I already did 120 epochs and it's still not great." means I don't have to ask how many images you have, because it doesn't matter. 

Plus, generating samples or a checkpoint midway through an iteration isn't really right either. If you save at step 105 and the optimizer steps at 106, then all the stuff you just trained doesn't go into the save or the samples.

It's simpler overall. I don't have to have six different fields of checkboxes to account for every possibility of how someone wants to arbitrarily count something.


Or, the TL/DR version:

No. Counting by steps is bad and pointless, and should go away entirely.


And class counts and stuff will be fixed with the merge I do later to night, as well as lots of other improvements.

**Comment by ekeric13** [comment]
Thanks for the response.

A step would just be how often you update the gradient. I agree that when you did a batch size of two and it showed double the number of steps that is unintuitive.

I honestly just did the math of (totalSteps * batchSize / numImages)  (never used the gradient accumulation).

I really think it's just a matter of how you do the math. 

Can you answer the "save model frequency by epoch" question though?
Still don't get what each integer is suppose to represent.

Will get some python 3.8 fixes in for you tonight as well.

**Comment by d8ahazard** [comment]
One step of the epoch is a full iteration of the set of instance images. If
you have 100 images, an epoch is 100 steps.

**Comment by ekeric13** [comment]
Sorry I should have been more specific. This UI is really confusing:

<img width="464" alt="Screenshot 2023-01-05 at 5 35 51 PM" src="https://user-images.githubusercontent.com/6489651/210911871-6a09d479-da84-4725-b7b4-dbf3b76459de.png">

I don't know how it determines how frequency to save preview images or checkpoints.

**Comment by d8ahazard** [comment]
One full iteration over your training dataset. If you have save preview
frequency at 1 and you have 100 images, it will save every 100 steps. 2 =
200. Etc.
